### PR TITLE
VEN-944 | Batch printing customer addresses on A4 papers

### PR DESCRIPTION
**Pull request by ahie**
## Description :sparkles:

Adds ability to generate sticker PDFs in the unmarked ws notice list page. Additionally refactors batch offer sending under the same list tool component.

### Closes :no_good_woman:

[VEN-944](https://helsinkisolutionoffice.atlassian.net/browse/VEN-944)

### Automated tests :gear:️

- Added `src/common/listActions/__test__/ListActions.test.tsx`
- Updated relevant tests and snapshots

### Manual testing :construction_worker_man:

- Open ws notices list page
- Select rows
- Choose "Tulosta tarrakirjeet" from "Toiminnot" dropdown
- If some of the selected notices do not have a sticker number, an error toast should appear when clicking the action button, otherwise it should be possible to save the generated pdf to disk or open it in the browser

### CODE REVIEW

File: src/features/unmarkedWsNoticeList/utils.ts [hunk 1]
```diff
@@ -93,3 +110,51 @@ export const getDraftedOffers = (applications: UnmarkedWinterStorageNotice[]) =>
       },
     ];
   }, []);
+
+export const generateAndSaveStickerPDF = (customers: CustomerInfo[]) => {
```
Comment (ahie) [inline src/features/unmarkedWsNoticeList/utils.ts:114]: First attempt to generate the pdf was to use jsPDF's html support and directly rendering a react component's markup. Due to several issues with the output (css units and typesetting being the main ones), the best way to achieve the desired output specified in the designs was to do this programmatically though the library's API.

File: src/common/listActions/ListActions.tsx [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+      <Select
+        onChange={(e) => setSelectedAction(listActions.find((action) => action.id === e.target.value))}
+        value={selectedAction ? selectedAction.id : ''}
+        options={options}
+        className={styles.select}
+      />
+      {selectedAction &&
+        selectedAction.renderComponent &&
+        selectedAction.renderComponent(selectedRows, resetSelectedRows)}
+      {selectedAction && selectedAction.onClick && (
```
Comment (Nizar-Rahme) [inline src/common/listActions/ListActions.tsx:47]:
```suggestion
      {selectedAction?.onClick && (
```

File: src/common/listActions/ListActions.tsx [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+};
+
+interface ListActionsProps<T> {
+  selectedRows: T[];
+  listActions: ListActionItem<T>[];
+  resetSelectedRows(): void;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const ListActions = <T extends any>({ selectedRows, resetSelectedRows, listActions }: ListActionsProps<T>) => {
```
Comment (Nizar-Rahme) [inline src/common/listActions/ListActions.tsx:24]: Could `any` be replaced with `unknown` or some `union` type?

File: src/common/listActions/ListActions.tsx [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    <div className={styles.container}>
+      <Select
+        onChange={(e) => setSelectedAction(listActions.find((action) => action.id === e.target.value))}
+        value={selectedAction ? selectedAction.id : ''}
+        options={options}
+        className={styles.select}
+      />
+      {selectedAction &&
+        selectedAction.renderComponent &&
+        selectedAction.renderComponent(selectedRows, resetSelectedRows)}
```
Comment (Nizar-Rahme) [inline src/common/listActions/ListActions.tsx:46]:
```suggestion
      {selectedAction?.renderComponent?.(selectedRows, resetSelectedRows)}
```

**Review by Nizar-Rahme — APPROVED** [review]
Overall it looks good to me! 
I left a couple of small comments, you might consider.

File: src/common/listActions/ListActions.tsx (hunk 2 above)
Comment (ahie) [reply src/common/listActions/ListActions.tsx]: Done.

File: src/common/listActions/ListActions.tsx (hunk 3 above)
Comment (ahie) [reply src/common/listActions/ListActions.tsx]: Done. Now it's "Object or any primitive type".

File: src/common/listActions/ListActions.tsx (hunk 4 above)
Comment (ahie) [reply src/common/listActions/ListActions.tsx]: Done.

**Pull request merged**